packages/flutter_tools/test/general.shard/fake_process_manager_test.dart

```diff
@@ -172,4 +175,149 @@ void main() {
       expect(stdout, 'stdout'.codeUnits);
     });
   });
+
+  group(FakeProcessManager, () {
+    late FakeProcessManager manager;
+
+    setUp(() {
+      manager = FakeProcessManager.empty();
+    });
+
+    group('start', () {
+      testWithoutContext('can run a fake command', () async {
+        manager.addCommand(const FakeCommand(command: <String>['faketool']));
+
+        final Process process = await manager.start(<String>['faketool']);
+        expect(await process.exitCode, 0);
+        expect(await utf8.decodeStream(process.stdout), isEmpty);
+        expect(await utf8.decodeStream(process.stderr), isEmpty);
+      });
+
+      testWithoutContext('outputFollowsExit delays stderr, stdout until after process exit', () async {
+        manager.addCommand(const FakeCommand(
+          command: <String>['faketool'],
+          stderr: 'hello',
+          stdout: 'world',
+          outputFollowsExit: true,
+        ));
+
+        final List<int> stderrBytes = <int>[];
+        final List<int> stdoutBytes = <int>[];
+
+        // Start the process.
+        final Process process = await manager.start(<String>['faketool']);
+        final StreamSubscription<List<int>> stderrSubscription = process.stderr.listen((List<int> chunk) { stderrBytes.addAll(chunk); });
+        final StreamSubscription<List<int>> stdoutSubscription = process.stdout.listen((List<int> chunk) { stdoutBytes.addAll(chunk); });
+
+        // Immediately after exit, no output is emitted.
+        await process.exitCode;
+        expect(utf8.decode(stderrBytes), isEmpty);
+        expect(utf8.decode(stdoutBytes), isEmpty);
+
+        // Output is emitted asynchronously after process exit.
+        await Future.wait(<Future<void>>[
+          stderrSubscription.asFuture(),
+          stdoutSubscription.asFuture(),
+        ]);
+        expect(utf8.decode(stderrBytes), 'hello');
+        expect(utf8.decode(stdoutBytes), 'world');
+
+        // Clean up stream subscriptions.
+        await stderrSubscription.cancel();
+        await stdoutSubscription.cancel();
+      });
+    });
+
+    group('run', () {
+      testWithoutContext('can run a fake command', () async {
+        manager.addCommand(const FakeCommand(command: <String>['faketool']));
+
+        final ProcessResult result = await manager.run(<String>['faketool']);
+        expect(result.exitCode, 0);
+        expect(result.stdout, isEmpty);
+        expect(result.stderr, isEmpty);
+      });
+
+      testWithoutContext('stderr, stdout are String if encoding is unspecified', () async {
+        manager.addCommand(const FakeCommand(command: <String>['faketool']));
+
+        final ProcessResult result = await manager.run(<String>['faketool']);
+        expect(result.exitCode, 0);
+        expect(result.stdout, isA<String>());
+        expect(result.stderr, isA<String>());
+      });
+
+      testWithoutContext('stderr, stdout are List<int> if encoding is null', () async {
```

Comment:
ouch

Reply:
Yeah I'd forgotten this was even a thing! Until https://github.com/flutter/flutter/pull/103947 we weren't doing the right thing in either `run` or `runSync` but in two completely different ways!